File: okf/src/okf/context.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from okf.bundle.document import OKFDocument
from okf.bundle.paths import path_to_concept_id


class ContextGenerator:
    """Generates prompt-ready context snippets for coding agents from OKF bundles."""

    def __init__(self, bundle_root: Path):
        self.bundle_root = Path(bundle_root).resolve()

    def get_summary_context(self) -> str:
        """Generates a concise progressive disclosure tree index of the bundle."""
        if not self.bundle_root.exists():
            return "No OKF bundle found."

        index_path = self.bundle_root / "index.md"
        if index_path.exists():
            index_content = index_path.read_text(encoding="utf-8")
        else:
            index_content = "Root index.md not found."

        output = [
            "# Project Knowledge Base (OKF)",
            f"**Bundle Location**: `{self.bundle_root.name}`",
            "",
            "## Progressive Disclosure Index",
            "",
            index_content,
            "",
            "---",
            "*(Agent Note: Use `okf get <concept_id>` or open specific `.md` files under `.okf/` for detailed schemas and architecture details.)*"
        ]
        return "\n".join(output)

    def get_full_concept_context(self, concept_id: str) -> str:
        """Fetch full markdown content of a specific concept."""
        file_path = self.bundle_root / f"{concept_id.lstrip('/')}.md"
        if not file_path.exists():
            return f"Concept '{concept_id}' not found in bundle."

        doc = OKFDocument.parse(file_path.read_text(encoding="utf-8"))
        fm = doc.frontmatter

        header = [
            f"# Concept: {fm.get('title', concept_id)}",
            f"- **Type**: {fm.get('type', 'Unknown')}",
            f"- **Description**: {fm.get('description', 'N/A')}",
        ]
        if fm.get("resource"):
            header.append(f"- **Resource**: `{fm.get('resource')}`")
        if fm.get("tags"):
            header.append(f"- **Tags**: {', '.join(fm.get('tags'))}")

        return "\n".join(header) + "\n\n" + doc.body

File: okf/context.go
package okf

import (
	"fmt"
	"os"
	"path/filepath"
	"slices"
	"strings"
)

// SummaryContext renders a progressive-disclosure prompt snippet from the
// bundle's root index.md, for use as agent context.
func SummaryContext(root string) (string, error) {
	info, err := os.Stat(root)
	if err != nil || !info.IsDir() {
		return "No OKF bundle found.", nil
	}

	indexContent := "Root index.md not found."
	if data, err := os.ReadFile(filepath.Join(root, indexFileName)); err == nil {
		indexContent = string(data)
	}

	lines := []string{
		"# Project Knowledge Base (OKF)",
		fmt.Sprintf("**Bundle Location**: `%s`", filepath.Base(root)),
		"",
		"## Progressive Disclosure Index",
		"",
		indexContent,
		"",
		"---",
		"*(Agent Note: Use `okf get <concept_id>` or open specific `.md` files under `.okf/` for detailed schemas and architecture details.)*",
	}
	return strings.Join(lines, "\n"), nil
}

// ConceptContext renders the full prompt-ready markdown for a single
// concept, including a metadata header derived from its frontmatter. If
// conceptID doesn't resolve to a file directly, it falls back to an exact,
// case-sensitive match against a concept's `aliases` frontmatter list
// (SPEC.md §3.1: "aliases ... Alternative lookup names").
func ConceptContext(root string, conceptID string) (string, error) {
	concept, found, err := resolveConcept(root, conceptID)
	if err != nil {
		return "", err
	}
	if !found {
		return fmt.Sprintf("Concept '%s' not found in bundle.", conceptID), nil
	}

	fm := concept.Doc.Frontmatter
	lines := []string{
		fmt.Sprintf("# Concept: %s", stringFrontmatterOr(fm["title"], concept.ID)),
		fmt.Sprintf("- **Type**: %s", stringFrontmatterOr(fm["type"], "Unknown")),
		fmt.Sprintf("- **Description**: %s", stringFrontmatterOr(fm["description"], "N/A")),
	}
	if resource := stringFrontmatterOr(fm["resource"], ""); resource != "" {
		lines = append(lines, fmt.Sprintf("- **Resource**: `%s`", resource))
	}
	if tags := frontmatterStringSlice(fm["tags"]); len(tags) > 0 {
		lines = append(lines, fmt.Sprintf("- **Tags**: %s", strings.Join(tags, ", ")))
	}
	if backlinks, err := Backlinks(root, concept.ID); err == nil && len(backlinks) > 0 {
		cited := make([]string, len(backlinks))
		for i, id := range backlinks {
			cited[i] = "`" + id + "`"
		}
		lines = append(lines, fmt.Sprintf("- **Cited by**: %s", strings.Join(cited, ", ")))
	}

	return strings.Join(lines, "\n") + "\n\n" + concept.Doc.Body, nil
}

// resolveConcept finds a concept by ID, trying the literal ID as a file
// path first. If no file exists there, it scans the bundle for a concept
// whose `aliases` frontmatter list contains conceptID as an exact,
// case-sensitive match. The bool return reports whether a concept was
// found at all.
func resolveConcept(root string, conceptID string) (ConceptFile, bool, error) {
	path := filepath.Join(root, strings.TrimLeft(conceptID, "/")+".md")
	if _, err := os.Stat(path); err == nil {
		data, err := os.ReadFile(path)
		if err != nil {
			return ConceptFile{}, false, err
		}
		doc, err := ParseDocument(string(data))
		if err != nil {
			return ConceptFile{}, false, err
		}
		return ConceptFile{ID: strings.TrimLeft(conceptID, "/"), Path: path, Doc: doc}, true, nil
	}

	concepts, err := walkConceptFiles(root)
	if err != nil {
		return ConceptFile{}, false, err
	}
	for _, c := range concepts {
		if c.Err == nil && slices.Contains(frontmatterStringSlice(c.Doc.Frontmatter["aliases"]), conceptID) {
			return c, true, nil
		}
	}
	return ConceptFile{}, false, nil
}

// frontmatterStringSlice extracts a []string from a frontmatter value,
// tolerating both the []string produced by normalizeYAMLValue and a raw
// []any with mixed element types.
func frontmatterStringSlice(v any) []string {
	switch val := v.(type) {
	case []string:
		return val
	case []any:
		out := make([]string, 0, len(val))
		for _, e := range val {
			if s, ok := e.(string); ok {
				out = append(out, s)
			} else {
				out = append(out, fmt.Sprint(e))
			}
		}
		return out
	default:
		return nil
	}
}
